From: John Ogness <john.ogness@linutronix.de>
Date: Tue, 26 Sep 2023 14:43:30 +0000
Subject: [PATCH 099/108] printk: nbcon: Add context to console_is_usable()
Origin: https://www.kernel.org/pub/linux/kernel/projects/rt/6.6/older/patches-6.6-rc6-rt10.tar.xz

The nbcon consoles have 2 callbacks to be used for different
contexts. In order to determine if an nbcon console is usable,
console_is_usable() needs to know if it is a context that
requires atomic support or is sufficient with threading
support.

Add an extra parameter @use_atomic to denote this.

Signed-off-by: John Ogness <john.ogness@linutronix.de>
Signed-off-by: Sebastian Andrzej Siewior <bigeasy@linutronix.de>
---
 kernel/printk/internal.h |   15 +++++++++------
 kernel/printk/nbcon.c    |    6 +++---
 kernel/printk/printk.c   |    6 ++++--
 3 files changed, 16 insertions(+), 11 deletions(-)

--- a/kernel/printk/internal.h
+++ b/kernel/printk/internal.h
@@ -86,7 +86,7 @@ void nbcon_kthread_create(struct console
  * which can also play a role in deciding if @con can be used to print
  * records.
  */
-static inline bool console_is_usable(struct console *con, short flags)
+static inline bool console_is_usable(struct console *con, short flags, bool use_atomic)
 {
 	if (!(flags & CON_ENABLED))
 		return false;
@@ -95,10 +95,13 @@ static inline bool console_is_usable(str
 		return false;
 
 	if (flags & CON_NBCON) {
-		if (!con->write_atomic)
-			return false;
-		if (!con->write_thread || !con->kthread)
-			return false;
+		if (use_atomic) {
+			if (!con->write_atomic)
+				return false;
+		} else {
+			if (!con->write_thread || !con->kthread)
+				return false;
+		}
 	} else {
 		if (!con->write)
 			return false;
@@ -161,7 +164,7 @@ static inline void nbcon_init(struct con
 static inline void nbcon_free(struct console *con) { }
 static bool nbcon_console_emit_next_record(struct console *con) { return false; }
 
-static inline bool console_is_usable(struct console *con, short flags) { return false; }
+static inline bool console_is_usable(struct console *con, short flags, bool use_atomic) { return false; }
 
 #endif /* CONFIG_PRINTK */
 
--- a/kernel/printk/nbcon.c
+++ b/kernel/printk/nbcon.c
@@ -997,7 +997,7 @@ static bool nbcon_kthread_should_wakeup(
 
 	cookie = console_srcu_read_lock();
 	flags = console_srcu_read_flags(con);
-	is_usable = console_is_usable(con, flags);
+	is_usable = console_is_usable(con, flags, false);
 	console_srcu_read_unlock(cookie);
 
 	if (!is_usable)
@@ -1069,7 +1069,7 @@ static int nbcon_kthread_func(void *__co
 
 		con_flags = console_srcu_read_flags(con);
 
-		if (console_is_usable(con, con_flags)) {
+		if (console_is_usable(con, con_flags, false)) {
 			/*
 			 * Ensure this stays on the CPU to make handover and
 			 * takeover possible.
@@ -1246,7 +1246,7 @@ static void __nbcon_atomic_flush_all(boo
 			if (!(flags & CON_NBCON))
 				continue;
 
-			if (!console_is_usable(con, flags))
+			if (!console_is_usable(con, flags, true))
 				continue;
 
 			memset(ctxt, 0, sizeof(*ctxt));
--- a/kernel/printk/printk.c
+++ b/kernel/printk/printk.c
@@ -2945,7 +2945,7 @@ static bool console_flush_all(bool do_co
 			if ((flags & CON_NBCON) && con->kthread)
 				continue;
 
-			if (!console_is_usable(con, flags))
+			if (!console_is_usable(con, flags, true))
 				continue;
 			any_usable = true;
 
@@ -3813,8 +3813,10 @@ static bool __pr_flush(struct console *c
 			 * that they make forward progress, so only increment
 			 * @diff for usable consoles.
 			 */
-			if (!console_is_usable(c, flags))
+			if (!console_is_usable(c, flags, true) &&
+			    !console_is_usable(c, flags, false)) {
 				continue;
+			}
 
 			if (flags & CON_NBCON) {
 				printk_seq = nbcon_seq_read(c);
